### ISSUE TITLE
torch::jit::script::Module::to(torch::kDouble) also casts buffers

### ISSUE THREAD

**Issue by IgnacioJPickering** [pytorch/pytorch#39421]
## 📚 Documentation

The C++ API documentation for ``torch::jit::script::Module::to(at::Device, at::ScalarType, bool)``
indicates that the function casts all Parameters to ``at::ScalarType``, however, the function also casts the Module's buffers. 

The documentation in question is in:
https://pytorch.org/cppdocs/api/structtorch_1_1jit_1_1_module.html#_CPPv4N5torch3jit6Module2toEN2at6DeviceEN2at10ScalarTypeEb

And you can see the behaviour of the function in: 
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/api/module.cpp

I'm not sure if this is a bug, a documentation issue, or I'm being a bit pedantic, so I'm raising it as a documentation issue, but I believe the behaviour of the python API is different. IIRC model.double() leaves buffers of type ``torch.long`` unchanged, but buffers of type ``torch::kLong`` are cast to ``torch::kDouble`` by to() using the C++ API.

**Comment by ailzhang** [comment]
According to eager documentation it's supposed to case `buffer` as well. And it matches eager behavior. 
```
In [8]: a.register_buffer('b', torch.rand(3,3))

In [9]: list(a.buffers())
Out[9]:
[tensor([[0.7092, 0.7046, 0.4389],
         [0.0908, 0.0132, 0.0891],
         [0.7526, 0.2759, 0.5418]])]

In [10]: b = a.to(torch.double)

In [12]: list(b.parameters())
Out[12]:
[Parameter containing:
 tensor([1., 1., 1.], dtype=torch.float64, requires_grad=True),
 Parameter containing:
 tensor([0., 0., 0.], dtype=torch.float64, requires_grad=True)]

In [13]: list(a.buffers())
Out[13]:
[tensor([[0.7092, 0.7046, 0.4389],
         [0.0908, 0.0132, 0.0891],
         [0.7526, 0.2759, 0.5418]], dtype=torch.float64)]
```

**Comment by IgnacioJPickering** [comment]
It is similar for eager, but not the same. 
In eager only floating point tensors are cast to double, but the C++ API casts also torch.long to double
this is the corresponding code in pytorch
```
def double(self):
        r"""Casts all floating point parameters and buffers to ``double`` datatype.

        Returns:
            Module: self
        """
        return self._apply(lambda t: t.double() if t.is_floating_point() else t)

```

**Comment by IgnacioJPickering** [comment]
You are right that it casts buffers in both, I had an issue because some of my buffers were cast to double only in C++ and I didn't understand why, and I tought only parameters were cast in pytorch, but it seems that both pytorch and libtorch cast parameters and buffers, but libtorch casts everything, pytorch only casts floating point parameters / buffers.

**Comment by gluefox** [comment]
Hey, this is still an open issue.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
